src/blockdata/witness.rs

```diff
@@ -0,0 +1,395 @@
+        let witness_elements = VarInt::consensus_decode(&mut d)?.0 as usize;
+        if witness_elements == 0 {
+            Ok(Witness::default())
+        } else {
+            let mut cursor = 0usize;
+            let mut last = 0usize;
+            let mut second_to_last = 0usize;
+
+            // this number should be determined as high enough to cover most witness, and low enough
+            // to avoid wasting space without reallocating
+            let mut content = vec![0u8; 128];
+
+            for _ in 0..witness_elements {
+                second_to_last = last;
+                last = cursor;
+                let element_size_varint = VarInt::consensus_decode(&mut d)?;
+                let element_size_varint_len = element_size_varint.len();
+                let element_size = element_size_varint.0 as usize;
+                let required_len = cursor
+                    .checked_add(element_size)
+                    .ok_or_else(|| self::Error::OversizedVectorAllocation {
+                        requested: usize::max_value(),
+                        max: MAX_VEC_SIZE,
+                    })?
+                    .checked_add(element_size_varint_len)
+                    .ok_or_else(|| self::Error::OversizedVectorAllocation {
+                        requested: usize::max_value(),
+                        max: MAX_VEC_SIZE,
+                    })?;
+
+                if required_len > MAX_VEC_SIZE {
+                    return Err(self::Error::OversizedVectorAllocation {
+                        requested: required_len,
+                        max: MAX_VEC_SIZE,
+                    });
+                }
+
+                resize_if_needed(&mut content, required_len);
+                element_size_varint
+                    .consensus_encode(&mut content[cursor..cursor + element_size_varint_len])?;
+                cursor += element_size_varint_len;
+                d.read_exact(&mut content[cursor..cursor + element_size])?;
+                cursor += element_size;
+            }
+            content.truncate(cursor);
+            Ok(Witness {
+                content,
+                witness_elements,
+                last,
+                second_to_last,
+            })
+        }
+    }
+}
+
+fn resize_if_needed(vec: &mut Vec<u8>, required_len: usize) {
+    if required_len >= vec.len() {
+        let mut new_len = vec.len().max(1);
+        while new_len <= required_len {
+            new_len *= 2;
+        }
+        vec.resize(new_len, 0);
+    }
+}
+
+impl Encodable for Witness {
+    fn consensus_encode<W: Write>(&self, mut writer: W) -> Result<usize, io::Error> {
+        let len = VarInt(self.witness_elements as u64);
+        len.consensus_encode(&mut writer)?;
+        writer.emit_slice(&self.content[..])?;
+        Ok(self.content.len() + len.len())
+    }
+}
+
+impl Witness {
+    /// Creates [`Witness`] object from an array of byte-arrays
+    pub fn from_vec(vec: Vec<Vec<u8>>) -> Self {
```

Comment:
Nit: why not `impl From<Vec<Vec<u8>> for Witness` instead?

And probably `pub fn from<T: AsRef<[u8]>>(bytes: &[T])` will be better.

Reply:
Because `From` is for cheap/trivial conversions. This is neither.

Reply:
> `From` is for cheap/trivial conversions

Not really, `impl From<&str> for String` exists for instance and that conversion is not usually considered "cheap".

@dr-orlovsky I consider `AsRef<[u8]>` a footgun since it would accept strings without explicit `as_bytes()`